### DISCUSSION TITLE
Add option to disable middleware persist

### DISCUSSION THREAD

**Discussion by prtmwrkr** [pmndrs/zustand#2572] [Ideas]
The title says it all - will add an example

```
import { create } from 'zustand'
import { persist, createJSONStorage } from 'zustand/middleware'

export const useBearStore = create(
  persist(
    (set, get) => ({
      bears: 0,
      addABear: () => set({ bears: get().bears + 1 }),
    }),
    {
      enabled: false, // (optional), if omitted is considered true
      name: 'food-storage',
      storage: createJSONStorage(() => sessionStorage),
    },
  ),
)
```

My use case is more complex than this one, I use `createStore` to create the store within a provider, receiving so different initialisation option (and the flag enabled would be one of this) so allowing me to have the same store working in two different modes in different pages.

**Comment by dai-shi** [comment]
In general, I'd prefer disabling the middleware entirely. 

```js
import { create } from 'zustand'
import { persist, createJSONStorage } from 'zustand/middleware'

export const useBearStore = create(
  (enabled ? persist : (fn) => fn)(
    (set, get) => ({
      bears: 0,
      addABear: () => set({ bears: get().bears + 1 }),
    }),
    {
      name: 'food-storage',
      storage: createJSONStorage(() => sessionStorage),
    },
  ),
)
```

But, it might be tricky and typing would be hard.
As `devtools` have `enabled` option, it might be reasonable to have it in `persist`. What do others think?
(Though, I'm not a big fan of options with default truthy value.)

**Reply by dai-shi** [reply]
(Or, could we make both `disabled` option?)

**Reply by prtmwrkr** [reply]
Disabling entirely was what I initially tried, but as you mentioned TS complained, and I didn't spend much time figuring out what was the issue. Having and `enabled`/`disabled` flag would make this much easier to obtain.
As a temporary workaround I'm using different `name` for case in which I want to persist the data (not optimal, but it works in my case).

**Reply by dai-shi** [reply]
Hm, can you please try this in the meantime?

```ts
import { create } from 'zustand'
import { persist, createJSONStorage } from 'zustand/middleware'

const myPersist: typeof persist = enabled ? persist : (fn: any) => any

export const useBearStore = create(
  myPersist(
    (set, get) => ({
      bears: 0,
      addABear: () => set({ bears: get().bears + 1 }),
    }),
    {
      name: 'food-storage',
      storage: createJSONStorage(() => sessionStorage),
    },
  ),
)
```

**Reply by prtmwrkr** [reply]
`const myPersist: typeof persist = enabled ? persist : (fn: any) => fn`

This works, and TS doesn't complain.
I can go with this as workaround, but imho having this exposed through the api would be a better option, even more considering other middlewares already do.

Thanks for the help here.

**Reply by riccardolardi** [reply]
Here's the TS if `any` is restricted:
```ts
const conditionalPersist = (
  persist ? _persist : (fn: unknown) => fn
) as typeof _persist;
```

**Comment by dai-shi** [comment]
RFC:

1. Should we add `enabled` option to `persist`?
2. Or should we add `disabled` option to `persist`?
3. Then what should we do with `devtools`?
4. Any other middleware which need such an option?
